src/pyhf/modifiers/shapesys.py

```diff
@@ -58,46 +64,39 @@ def __init__(self, shapesys_mods, pdfconfig, mega_mods):
             ]
         )
 
-        if self._shapesys_indices:
-            access_rows = []
-            shapesys_mask = default_backend.astensor(self._shapesys_mask)
-            for mask, inds in zip(shapesys_mask, self._shapesys_indices):
-                summed_mask = default_backend.sum(mask[:, 0, :], axis=0)
-                assert default_backend.shape(
-                    summed_mask[summed_mask > 0]
-                ) == default_backend.shape(default_backend.astensor(inds))
-                # make masks of > 0 and == 0
-                positive_mask = summed_mask > 0
-                zero_mask = summed_mask == 0
-                # then apply the mask
-                summed_mask[positive_mask] = inds
-                summed_mask[zero_mask] = len(self._parindices) - 1
-                access_rows.append(summed_mask.tolist())
-            self._factor_access_indices = default_backend.tolist(
-                default_backend.stack(access_rows)
-            )
-            self.finalize(pdfconfig)
-        else:
-            self._factor_access_indices = None
+        self.finalize(pdfconfig)
+
+        global_concatenated_bin_indices = [
+            [[j for c in pdfconfig.channels for j in range(pdfconfig.channel_nbins[c])]]
+        ]
+
+        self._access_field = default_backend.tile(
+            global_concatenated_bin_indices, (len(pnames), self.batch_size or 1, 1)
+        )
+        # access field is shape (sys, batch, globalbin)
+        for s, syst_access in enumerate(self._access_field):
+            for t, batch_access in enumerate(syst_access):
+                selection = self.parameters_helper.index_selection[s][t]
+                for b, bin_access in enumerate(batch_access):
+                    self._access_field[s, t, b] = (
+                        selection[bin_access] if bin_access < len(selection) else 0
+                    )
 
         self._precompute()
         events.subscribe('tensorlib_changed')(self._precompute)
 
     def _precompute(self):
         tensorlib, _ = get_backend()
+        if not self.parameters_helper.index_selection:
+            return
         self.shapesys_mask = tensorlib.astensor(self._shapesys_mask)
+        self.shapesys_mask = tensorlib.tile(
+            self.shapesys_mask, (1, 1, self.batch_size or 1, 1)
```

Comment:
why is the `batch_size` in that dimension?

Reply:
because the alphasets shape is that dimension.